ts/src/header-validator/validate-json.ts

```diff
@@ -725,6 +726,101 @@ const priorityField: StructFields<Priority> = {
   priority: field('priority', int64, 0n),
 }
 
+function isDebugTypeSupported(ctx: RegistrationContext, s: string) {
+  return ctx.aggregatableDebugTypes.includes(s)
+}
+
+function aggregatableDebugType(
+  ctx: RegistrationContext,
+  j: Json
+): Maybe<string> {
+  return string(ctx, j).peek((s) => {
+    if (!isDebugTypeSupported(ctx, s)) {
+      ctx.warning('unknown type')
+    }
+  })
+}
+
+export type KeyPiece = {
+  keyPiece: bigint
+}
+
+const keyPieceField: StructFields<KeyPiece> = {
+  keyPiece: field('key_piece', hex128),
+}
+
+export type AggregatableDebugReportingData = KeyPiece & {
+  types: string[]
+  value: number
+}
+
+function aggregatableDebugReportingData(
+  ctx: RegistrationContext,
+  j: Json
+): Maybe<AggregatableDebugReportingData> {
+  return struct(ctx, j, {
+    types: field('types', (ctx, j) =>
+      array(ctx, j, aggregatableDebugType, { minLength: 1 })
+    ),
+    value: field('value', aggregatableValue),
+
+    ...keyPieceField,
+  })
+}
+
+function aggregatableDebugReportingDataList(
+  ctx: RegistrationContext,
+  j: Json
+): Maybe<AggregatableDebugReportingData[]> {
+  return array(ctx, j, aggregatableDebugReportingData).filter((data) => {
+    const types = new Set<string>()
+    const dups = new Set<string>()
+    for (const d of data) {
+      for (const t of d.types) {
+        if (!isDebugTypeSupported(ctx, t)) {
+          continue
+        }
+        if (types.has(t)) {
+          dups.add(t)
+        } else {
+          types.add(t)
+        }
+      }
+    }
+    if (dups.size > 0) {
+      ctx.error(`duplicate type: ${Array.from(dups).join(', ')}`)
+      return false
+    }
+    return true
+  })
+}
+
+function aggregationCoordinatorOriginField(
+  ctx: RegistrationContext,
+  j: JsonDict
+): Maybe<string> {
+  return field(
+    'aggregation_coordinator_origin',
+    aggregationCoordinatorOrigin,
+    ctx.vsv.aggregationCoordinatorOrigins[0]
+  )(ctx, j)
+}
```

Comment:
Instead of making this a function that immediately invokes the function returned by `field`, can we make this work more similarly to https://github.com/WICG/attribution-reporting-api/blob/69bed9def9b95eb796d8abf9c65943b374b60d28/ts/src/header-validator/validate-json.ts#L704-L710?

Then `Trigger` and `AggregatableDebugReportingConfig` could both extend an `AggregationCoordinatorOrigin` type.

Reply:
I thought about that as well, but I don't know what's the best way to set the default which needs ctx. Any suggestions?

Reply:
I can take a look at cleaning this up in a followup field.